File: Python/src/day_2.py
from src import utils

class Day2:
    def __init__(self):
        self.file_contents = utils.read_file ("../files/day_2.txt")
        self.levels = self.parse_content (self.file_contents)

    @staticmethod
    def parse_content(contents):
        lines = contents.splitlines();
        
        levels = []

        for l in lines:
            numbers = l.split(' ')
            converted_numbers = [int(numeric_string) for numeric_string in numbers]
            levels.append(converted_numbers)

        return levels
    
    @staticmethod
    def sequence_passed(level: list):

        is_ascending = level[0] < level[1]

        for i in range(1, len(level)):
            diff = level[i] - level[i - 1]

            if diff == 0 or diff > 3 or diff < -3:
                return False
            
            if is_ascending and level[i] < level[i - 1]:
                return False
            
            if not is_ascending and level[i] > level[i - 1]:
                return False
            
        return True
    
    @staticmethod
    def tryWithoutNumber(level: list):
        for i in range(len(level)):
            altered_numbers = level.copy()
            altered_numbers.pop(i)
            if Day2.sequence_passed (altered_numbers): 
                return True

        return False
    
    def run_part_1(self):
        passed_levels = list(filter(self.sequence_passed, self.levels))
        print(len(passed_levels))

    def run_part_2(self):
        passed_levels = []
        failed_levels = []
        
        for level in self.levels:
            if self.sequence_passed(level):
                passed_levels.append (level)
            else:
                failed_levels.append (level)

        # check through failed levels
        for level in failed_levels:
            if self.tryWithoutNumber(level):
                passed_levels.append(level)

        print(len(passed_levels))

File: cpp/Day2.cpp
#include "Day2.h"
#include "Utils.h"

#include <cstdio>
#include <string>
#include <iostream>
#include <sstream>
#include <queue>

Day2::Day2()
{
    std::cout << "Day 2" << std::endl;

    std::filesystem::path contentPath = "/Users/reubenduckering/Documents/Personal Repo/Advent-Of-Code-2024/files/day_2.txt";
    const auto lines = Utils::getContentLines (contentPath);

    for (const auto& line : lines)
    {
        std::istringstream stream (line);
        std::string l;

        vector<int> numbers;

        while (getline (stream, l, ' '))
            numbers.push_back (std::stoi (l));

        levels.push_back (numbers);
    }
}

void Day2::runPart1()
{
    int safeLevels = 0;

    for (const auto& level : levels)
        safeLevels += static_cast<int> (sequencePassed (level));

    std::cout << "Safe levels: " << safeLevels << std::endl;
}

void Day2::runPart2()
{
    int safeLevels = 0;
    int index = 0;

    std::queue<int> failedLevels;

    for (const auto& level : levels)
    {
        if (sequencePassed (level))
            ++safeLevels;
        else
            failedLevels.push (index);

        ++index;
    }

    // process failed levels

    auto tryWithoutNumbers = [this] (const std::vector<int>& level) {
        const size_t size = level.size();

        for (int i = 0; i < size; ++i)
        {
            std::vector<int> alteredLevels = level;
            alteredLevels.erase (alteredLevels.begin() + i);

            if (sequencePassed (alteredLevels))
                return true;
        }

        return false;
    };

    for (; ! failedLevels.empty(); failedLevels.pop())
    {
        const int i = failedLevels.front();

        const auto level = levels[i];

        if (tryWithoutNumbers (level))
            ++safeLevels;
    }

    std::cout << "Safe levels: " << safeLevels << std::endl;
}

bool Day2::sequencePassed (const vector<int>& level)
{
    const size_t size = level.size();

    bool isAscending = level[0] < level[1];

    for (int i = 1; i < size; ++i)
    {
        int diff = level[i - 1] - level[i];

        if (diff == 0)
            return false;

        if (diff > 3 || diff < -3)
            return false;

        if (isAscending && level[i] < level[i - 1])
            return false;

        if (! isAscending && level[i] > level[i - 1])
            return false;
    }

    return true;
}